### ISSUE TITLE
[PA1] Cyclic Dependency 

### ISSUE THREAD

**Issue by qwertyKey-s** [nus-cs2030/2223-s1#150]
#### How to avoid cyclic dependency?

![pic3](https://user-images.githubusercontent.com/114850052/193442314-35e820a5-afad-4523-aa13-a65eebb17f00.png)

From what I understand,
1. All `Term` are able to execute `.snd(Host)` which returns a `Host`
2. Only `Host` that have received the `.snd(Host)` signal are able to execute `.rcv()` which returns a `Term`
3. Only `Term` that have received the `.snd(Host).rcv()` signal are able to execute `.ack()` which returns a `Host`

My train of thought is,

`Term.snd(Host)` returns a `Host` --> `Term.snd(Host)` is a `Host` --> `Term` is dependent on `Host`

`Term.snd(Host).rcv()` returns a `Term` --> since `Term.snd(Host)` is a `Host` --> `Term.snd(Host).rcv()` returns a `Term` can be simplistically understood as `Host.rcv()` returns a `Term` --> `Host.rcv()` is a `Term` --> `Host` is dependent on `Term`

`Term` is dependent on `Host` && `Host` is dependent on `Term` --> cyclic dependency exists

**Comment by wilsonwid** [comment]
I initially faced this issue, but I think by reading between the lines the question implicitly asks you to return a type of `Host` instead of the `Host` itself. However, I do agree that slightly more information could have been provided in this respect to not confusion regarding this during the PA :(

**Comment by zhanyi3** [comment]
I also faced this issue when submitting to codecrunch. I managed to resolve it by only passing whatever information is required to the new Host instead of passing the whole Host object. Not sure if it helps but it became clearer to me where the cyclic dependency problem arises when I did a rough sketch of the reference pointers.

**Comment by RezwanAhmed123** [comment]
As @zhanyi3 said earlier the best way to figure out where the circular dependencies are originating from is to draw a sketch of all ur classes and their relation to each other (eg A points to B if it depends on B etc and usually depending means like inheriting from B or like it creates/involves B in its code or something) then from there you can figure out how to change your classes to prevent the circular dependency

**Comment by kingofburg** [comment]
Yes I agree from the above comments, it helped me when I draw out all the possible classes. Also, the error message given in code crunch showed me where the cyclicity was, and thus I managed to eliminate those using the error message! hope this helps

**Comment by eLTehh** [comment]
Hi, to add on:

For my "rcv" instance of a pager, I had it inherit from Term instead of Pager to avoid cyclic dependency. Hope this helps someone

**Comment by acekhoon** [comment]
Initially I also got cyclic dependency because 1. I did not return specific type which is subclass of Host (it should not be like "Host rcv()"), and 2. I used attirbutes which are not Term and Host (abstract class does not raise cyclic dependency error if you are using it as an attribute), which are quite implicit in this question. Check them in order to avoid cyclic dependency.
